src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/TypeNamesShouldNotMatchNamespacesTests.cs

```diff
@@ -9,12 +12,98 @@ public class TypeNamesShouldNotMatchNamespacesTests : DiagnosticAnalyzerTestBase
+        }
+
+        private static DiagnosticResult CSharpSystemResultAt(int line, int column, string typeName, string namespaceName)
+        {
+            return GetCSharpResultAt(line, column, TypeNamesShouldNotMatchNamespacesAnalyzer.SystemRule, typeName, namespaceName);
+        }
+
+        private static DiagnosticResult BasicDefaultResultAt(int line, int column, string typeName, string namespaceName)
+        {
+            return GetBasicResultAt(line, column, TypeNamesShouldNotMatchNamespacesAnalyzer.DefaultRule, typeName, namespaceName);
+        }
+
+        private static DiagnosticResult BasicSystemResultAt(int line, int column, string typeName, string namespaceName)
+        {
+            return GetBasicResultAt(line, column, TypeNamesShouldNotMatchNamespacesAnalyzer.SystemRule, typeName, namespaceName);
+        }
+
+        [Fact]
+        public void CA1724CSharpValidName()
+        {
+            VerifyCSharp(@"
+public class C
+{
+}");
+        }
+
+        [Fact]
+        public void CA1724CSharpInvalidNameMatchingFormsNamespaceInSystemRule()
+        {
+            VerifyCSharp(@"
+public class Forms
+{
+}",
+        CSharpSystemResultAt(2, 14, "Forms", "System.Windows.Forms"));
+        }
+
+        [Fact]
+        public void CA1724CSharpInvalidNameMatchingSdkNamespaceInDefaultRule()
+        {
+            var source = @"
+public class Sdk
+{
+}
+";
+            Document document = CreateDocument(source, LanguageNames.CSharp);
+            Project project = document.Project.AddMetadataReference(MetadataReference.CreateFromFile(typeof(Xunit.Sdk.AllException).Assembly.Location));
+            DiagnosticAnalyzer analyzer = GetCSharpDiagnosticAnalyzer();
+            GetSortedDiagnostics(analyzer, project.Documents.Single())
+                .Verify(analyzer,
+                        CSharpDefaultResultAt(2, 14, "Sdk", "Xunit.Sdk"));
+        }
+
+        [Fact]
+        public void CA1724VisualBasicValidName()
+        {
+            VerifyBasic(@"
+Public Class C
+End Class");
+        }
+
+        [Fact]
+        public void CA1724VisualBasicInvalidNameMatchingFormsNamespaceInSystemRule()
+        {
+            VerifyBasic(@"
+Public Class Forms
+End Class",
+        BasicSystemResultAt(2, 14, "Forms", "System.Windows.Forms"));
+        }
+
+        [Fact]
+        public void CA1724VisualBasicInvalidNameMatchingSdkNamespaceInDefaultRule()
+        {
+            var source = @"
+Public Class Sdk
+End Class";
+            Document document = CreateDocument(source, LanguageNames.VisualBasic);
+            Project project = document.Project.AddMetadataReference(MetadataReference.CreateFromFile(typeof(Xunit.Sdk.AllException).Assembly.Location));
+            DiagnosticAnalyzer analyzer = GetCSharpDiagnosticAnalyzer();
+            GetSortedDiagnostics(analyzer, project.Documents.Single())
+                .Verify(analyzer,
+                        BasicDefaultResultAt(2, 14, "Sdk", "Xunit.Sdk"));
```

Comment:
Since you took care to ensure the stability of the result in the case where components are duplicated, add a unit test for this. Perhaps

```
namespace A.B
{
}

namespace C.A
{
}

namespace D
{
    public class A {}
}
```

... should always complain about `A.B`.